release-25.4: sqlstats: fix unbounded memory growth bug in sql stats

kyle-a-wong opened the pull request:
Backport:
  * 1/1 commits from "sqlstats: flush ingester buffer if above max statements per txn" (#158527)
  * 1/1 commits from "sqlstats: change force flush event to be backport friendly" (#159621)

This PR also includes a commit to increase the default value of `sql.metrics.transaction_details.max_statement_stats` to `math.maxInt64` to disable this feature by default. Customers can set this setting to a lower value if they are experiencing SQL Stats related memory issues related to the issue fixed in this backport. 

Please see individual PRs for details.

/cc @cockroachdb/release

Release Justification: Fixes a critical bug seen by multiple customers that causes unexpected unbounded memory growth due to the SQL Stats Ingester.

kyle-a-wong committed:
sqlstats: flush ingester buffer if above max statements per txn

There have been a couple incidents recently where workloads running
multi day / million statement txns cause sql stats to have unbounded
memory growth. This is happening because SQL Stats are buffered by
sessionID until a transaction commits, which causes the buffer to
flush and the statement stats to be recorded in the SQL stats
subsystem. When there is a long running transaction with many
statements, we buffer is never flushed and continues to grow.

To fix this, we have introduced a new "flush" event that
forces the SQL stats ingester to flush sql stats in the current
session, if a certain threshold is met. This threshold is currently
set to the value of a new cluster setting:
`sql.metrics.transaction_details.max_statement_stats`.

If this threshold is met, the stats are automatically flushed. The side effect
of this is that these stats will not have an associated transaction
fingerprint id. This is because the transaction fingerprint id
isn't finalized until a transaction is complete and we don't know
the transaction fingerprint id at the time of this forced flush.
There is also no way to "backfill" or set the transaction
fingerprint id once the statement stat has been recorded to the
SQL Stats subsystem.

This commit also includes a change to the
`SQLStatsIngester.flushBuffer` method to recording transaction
stats when there are no statements stats in the buffer. This is
the case when the buffer is being force flushed in the scenario
mentioned above.

Fixes: #158800
Epic: None
Release note (bug fix): Addresses a bug with unbounded memory
growth when long running transactions are contain many statements.
In this case, the SQL Stats system will automatically flush these
before the transaction commits. The side effect of this is that
these statement statistics will no longer have an associated
transaction fingerprint id.

kyle-a-wong committed:
sqlstats: change force flush event to be backport friendly

PR #158527 introduced a fix to sql stats for a bug
that could cause unbounded memory growth. The original
implementation of this change changed the underlying
`flushBuffer` function that is used in the normal happy
path of sql stats ingestion. In order to make this change
more safe for backporting, the implementation of this
fix is being changed such that the existing flush
logic is the same and only the "forceFlush" path
has new logic.

Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/159646)
<!-- Reviewable:end -->

kyle-a-wong committed:
sqlstats: disable sql stats ingester force flush by default

A new cluster setting, `sql.metrics.transaction_details.max_statement_stats`,
was introduced to control the point where sql stats force flushes the
statement statistics buffer in the sql stats ingester. To limit impact
of the backport of this feature, the default of this value is being
set to the max int64 value, effectively disabling it.

Epic: None
Release note: None
